### PR TITLE
CPS: wait for the kernel to set up KNI interfaces

### PR DESCRIPTION
**Pull request by AltraMayor**
There is a race condition between the kernel and Gatekeeper that occasionally terminates the initialization of Gatekeeper with
the following log entries:

```
GATEKEEPER CPS: mnl_cb_run: cannot bring KNI kni_front up: No such device
GATEKEEPER CPS: Failed to configure KNI link on the front iface
```

OR

```
GATEKEEPER CPS: mnl_cb_run: cannot bring KNI kni_back up: No such device
GATEKEEPER CPS: Failed to configure KNI link on the back iface
```

This patch waits for the kernel to assign interface indexes to the KNI interfaces before setting the KNI interfaces to avoid
the race condition.

### CODE REVIEW

File: cps/main.c [hunk 1]
```diff
@@ -676,6 +676,35 @@ assign_cps_queue_ids(struct cps_config *cps_conf)
...
+			/* Interface is available. */
+			return 0;
+		}
+
+		i++;
+		sleep(1);
+	/*
+	 * Use greater than in the test below to account for the fact that
+	 * there is no wait before the first call of if_nametoindex().
+	 */
```
Comment (mengxiang0811) [inline cps/main.c:700]: Style: need to align with the above statements.

**Review by mengxiang0811 — CHANGES_REQUESTED** [review]
**CPS: wait for the kernel to set up KNI interfaces**

File: cps/main.c [hunk 2]
```diff
@@ -676,6 +676,35 @@ assign_cps_queue_ids(struct cps_config *cps_conf)
...
+	} while (i > wait_trials);
```
Comment (mengxiang0811) [inline cps/main.c:701]: If `wait_trials` is greater than `0`, then this condition will always be `false` after the first call of `if_nametoindex()`.

**Comment by AltraMayor** [comment]
Pull request ready for another review.

**Comment by AltraMayor** [comment]
The root cause of this issue has been found: `systemd` was renaming the KNI interfaces. `systemd` may be configured to rename an interface based **only** on its MAC address. This is done in production to have friendly interface names to be used in scripts. However, since a KNI interface in Gatekeeper has the same MAC address as its corresponding physical interface, this creates many problems.

The solution to use the PCI addresses of physical interfaces instead of their MAC addresses to identify them for `systemd`.

**Pull request closed without merging**